### PR TITLE
Fix citizen limit calculation

### LINKED ISSUE #3463: Max citizen number not increasing
### Minecolonies version 

Version: minecolonies-1.12.2-0.10.202-RELEASE-universal.jar

### Expected behavior

I have been attempting to raise my citizen cap, but it's simply not going up. I've built several houses (4 level 3 houses and 1 level 2), but the cap was stuck at 4. I then killed one of my citizens to see if this would work (I'm not a smart guy), and now the citizen cap is simply 3/3. I would expect the number to go up according to my housing levels (which if my math serves, should be 14.)

### Actual behaviour

The cap stays the exact same.

### Steps to reproduce the problem

Well, as you can see I've built a townhall and several houses. I will say; while messing around with the reposition feature, I accidentally broke my townhall and had to place a new one, I have no idea if this is somehow messing with the mod, but I didn't even think about it until some time later.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/70910296-max-citizen-number-not-increasing?utm_campaign=plugin&utm_content=tracker%2F43175463&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F43175463&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #4191: Firing a knight/ranger from barracks results in funny colonist count
### Minecolonies version

Version: minecolonies-0.10.493-RELEASE-universal

### Expected behavior

- Total colonist count to count all colonists correctly: 

   current colonists/total possible

### Actual behaviour

- The total possible is less then the current population. I've seen this issue prevent children from showing up if there was an extra citizen home available. Building another or upgrading a home would fix the ratio and generally 2 children would soon show up.

### PR DESCRIPTION
**Pull request by someaddons**
Closes #4191
Closes #3463

# Changes proposed in this pull request:
Now Calculates on building upgrade/destruction.
Recalculates after colony load, instead of reading old values.
Recruitment of additional guards over the citizen hut limit now works.
Level 0 guard-buildings no longer count towards the limit.

**Update**: Fixed another case and changed it to global instead. So that in the future you only need to adjust the maxCitizenCalc for new buildings/other stuff that houses.
The max citizens are now recalculated on:
- Assign and removal from a building
- Upgrade and destruction from a building
- removal of a citizen
- Colony load

Review please

### CODE REVIEW

File: src/api/java/com/minecolonies/api/tileentities/TileEntityRack.java [hunk 1]
```diff
@@ -340,15 +340,7 @@ public void read(final CompoundNBT compound)
...
-                inventory = new RackInventory(DEFAULT_SIZE + size * SLOT_PER_LINE)
-                {
-                    @Override
-                    protected void onContentsChanged(final int slot)
-                    {
-                        updateItemStorage();
-                        super.onContentsChanged(slot);
-                    }
-                };
+                inventory = new RackInventory(DEFAULT_SIZE + size * SLOT_PER_LINE);
```
Comment (someaddons) [inline src/api/java/com/minecolonies/api/tileentities/TileEntityRack.java:343]: our RackInventory class already has that overwrite, onContentsChanged was overwritten twice on accident

Comment (Raycoms) [reply src/api/java/com/minecolonies/api/tileentities/TileEntityRack.java:343]: ohh

File: src/main/java/com/minecolonies/coremod/colony/managers/CitizenManager.java [hunk 2]
```diff
@@ -92,9 +92,6 @@ public CitizenManager(final Colony colony)
     @Override
     public void read(@NotNull final CompoundNBT compound)
     {
-        maxCitizens = compound.getInt(TAG_MAX_CITIZENS);
-        potentialMaxCitizens = compound.getInt(TAG_POTENTIAL_MAX_CITIZENS);
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/colony/managers/CitizenManager.java:96]: Shouldn't we then recalc after read?

Comment (someaddons) [reply src/main/java/com/minecolonies/coremod/colony/managers/CitizenManager.java:96]: ye but we need the buildings to be loaded for that calculation, thats why it is added after the building manager read in the colony

Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/managers/CitizenManager.java:96]: ahkk

**Review by Raycoms — DISMISSED** [review]

File: src/api/java/com/minecolonies/api/tileentities/TileEntityRack.java (hunk 1 above)
Comment (Asherslab) [reply src/api/java/com/minecolonies/api/tileentities/TileEntityRack.java:343]: Was just checking that, glad to hear

**Review by Asherslab — DISMISSED** [review]

**Comment by marvin-bitterlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=4803568&bid=15983747)

**Review by marchermans — APPROVED** [review]

**Pull request merged**